updated user model according to comments

Nandu306 opened the pull request:
@emcoding

I've made most of the suggested changes, expect for the null:false constraint on the name field. Figuring out how to add that to the name field.

emcoding commented on app/models/assignment.rb:
```diff
@@ -12,28 +11,24 @@ class Assignment < ApplicationRecord
-    self.completed_assignments.each do |completed_assignment|
+    completed_assignments.each do |completed_assignment|
       total += completed_assignment.score
     end
```
O wait, that's a different model, I'm going to ignore things outside user.rb ;-)

Nandu306 replied:
Since this was a quick fix, I have now changed this as well.

emcoding reviewed (commented):
Hi Nandu,

So, how do you like your code now? Does it look better, nicer or... to you? 

One general comment, for future code review requests.
For every developer, reading code, especially someone else's code, is hard. 
To make it easier for other devs to review your code, you can do several things:
- In your PR, do one thing only (for example, in this case I reviewed the User.rb only; so it's good practice to open a PR with changes to User.rb only. And of course, other code that is changed because of the changes in User.rb, if any.)
- In the PR description, explain in more detail why you did what you did. (Tom, from ams.rb did a talk about that. I'll see if he can send a link.)
- That said, don't be afraid if things doesn't go right. If you pushed some files that didn't belong in the PR, because of git, or branches or whatever, just tell in the description why they are there. It's not a big deal. 
- The most important thing is that it is great that you ask for feedback. Well done. 

And happy coding :-)

emcoding commented on db/migrate/20220614122827_remove_null_from_user.rb:
```diff
@@ -0,0 +1,5 @@
+class RemoveNullFromUser < ActiveRecord::Migration[6.0]
+  def change
+    remove_column :users, :null, :boolean
```
https://guides.rubyonrails.org/active_record_migrations.html#changing-columns may help

Nandu306 commented:
Thanks Maud! I'm definitely looking at the code with a different pair of eyes and it's definitely useful for another person to give feedback. Since I reordered the user model, I thought I'd do the same with the other models as well which is why some of the other files were changed. Apologies.

emcoding commented on db/migrate/20220614163343_change_name_on_users.rb:
```diff
@@ -0,0 +1,5 @@
+  def change
+    change_column :users, :name, :string, null: false
+  end
+end
```
Adding a `null constraint` to an existing column , like in this migration, may make troubles in the future. 
Here's nicely explained why: https://testsuite.io/manage-database-null-contraints
That's why Rails offers the `change_column_null`-migration (see the link to the guides in my previous comment). 

So, consider to practice the column_null migration in this project. Your call, of course.

emcoding reviewed (commented):
Looks nice! Congrats.   
Think about if you want to work on the migration, but if you just take away the caveats and remember the null migration for next time, that is totally okay too.

emcoding commented:
Okay, now I am stalking you about migrations. Sorry! I forgot to mention the most easy way to add a null constraint, for future cases: After generating the migration, just edit the migration file before running `db:migrate`, and add `null: false` to the line where it's needed. 
For example: 

```
class CreateUsers < ActiveRecord::Migration[7.0]
  def change
    create_table :users do |t|
      t.string :thing, null: false  #manually type `null: false` in the file 
      t.integer :role, default: 0

      t.timestamps
    end
  end
end
```
And if you don't mind losing the data, you could rollback the user migration, add the  `null false` manually, and run db:migrate again.

Nandu306 commented:
> Okay, now I am stalking you about migrations. Sorry! I forgot to mention the most easy way to add a null constraint, for future cases: After generating the migration, just edit the migration file before running `db:migrate`, and add `null: false` to the line where it's needed. For example:
> 
> ```
> class CreateUsers < ActiveRecord::Migration[7.0]
>   def change
>     create_table :users do |t|
>       t.string :thing, null: false  #manually type `null: false` in the file 
>       t.integer :role, default: 0
> 
>       t.timestamps
>     end
>   end
> end
> ```
> 
> And if you don't mind losing the data, you could rollback the user migration, add the `null false` manually, and run db:migrate again.

I dropped the db, added change_column_null to one of the earlier user migrations files and then did db:create migrate.

Nandu306 commented:
Quick question: I've used the enums alongside if/else statements to control the views (a teacher would see/do different things compared to a student). Is that a smart way of handling things?

emcoding commented:
What matters is:  is it easy to follow (by your future self, and other devs), is it easy to make changes, and can it go wrong. 
Let's say there is a scale from simple to complex. The simplest is a boolean (for example: `can_add_score`) in the database, so you can check if a `user.can_add_acore?` A complex solution would be a Role model and Pundit or Cancancan for Role based authoristation.
I myself like to start with a more simple solution first. So, if there's only one thing that differs between roles, (like: `can_view_dashboard`) I would do the boolean attr in the database. 
Anything more complicated, I do the enum roles. 
With enum roles, I'd rather create methods in the User model like `can create assignments?`  and do the teacher/student check there.
That is a bit more explicit in the views.  
I live by The Rule of Three: if I had more than 3 roles, the enums is probably getting complex fast, so that would be a good moment to either extract a Role model, or go to Cancancan/Pundit.
Same: if I had more than 3 of the  `can_do_this`  methods, it's time for something else. 

And another consideration is in the views. Is it still manageable with the if/else conditions? Or is it getting confusing? Is it still as easy to add some feature in the view, or does it take a lot of time to figure out all the forks? 
Maybe it's easier to fix  with something like   rendering different partials for students and teachers. Because with Pundit and Cancancan you still have to do some if/else checks in the views. 

In general: yes, (starting with) enums is a good idea. Until it isn't anymore. That's mainly up to you to decide. Or play around with a different solution just for fun, and to learn what the pro's and con's of each solution is.
   
 Does this help?

Nandu306 commented:
Oops, thought I had replied to this. Yes, that makes sense. I suppose enum is probably the best choice for the app in question.
